V2 migration guide updates (#6006)
pydantic/pydantic: docs/migration.md

We recommend using a virtual environment to isolate your testing environment:

```bash
pip install --pre -U "pydantic>=2.0b1"
```

!!! warning "Beta release!"
    Note that there are still some rough edges, and while trying out the Pydantic V2 beta releases you may experience errors.

    We encourage you to try out the beta releases in a test environment and not in production.
    While we do not currently intend to make changes to the API, it's still possible that we
    will discover problems requiring breaking changes.

If you do encounter any issues, please [create an issue in GitHub](https://github.com/pydantic/pydantic/issues) using
the `bug V2` label. This will help us to actively monitor and track errors, and to continue to improve the library's
performance.

## Migration guide

### Changes to `pydantic.BaseModel`

Various method names have been changed; all non-deprecated `BaseModel` methods now have names matching either the
format `model_.*` or `__.*pydantic.*__`. Where possible, we have retained the deprecated methods with their old names
to help ease migration, but calling them will emit `DeprecationWarning`s.

> | Pydantic V2  | Pydantic V1 |

Reviewer: @dmontagu sanity check on this would be helpful, if you know what maps off the top of your head

description: Migrating from Pydantic V1.
---

The Pydantic V2 beta introduces a number of changes to the API, including some breaking changes.

We believe Pydantic V2 is ready for your production applications. This page provides a guide highlighting the most
important changes to help you migrate your code from Pydantic V1 to Pydantic V2.

## Install Pydantic V2 beta

Your feedback will be a critical part of ensuring that we have made the right tradeoffs with the API changes in
Pydantic V2.

To get started with the Pydantic V2 beta, install it from PyPI.
We recommend using a virtual environment to isolate your testing environment:

```bash

> pip install --pre -U "pydantic>=2.0b1"

Reviewer 2: Should we generate this from our plugin? Something like `{{ pydantic_version }}`

Reviewer: This would be good. Let's look into making this possible

```

!!! warning "Beta release!"
    Note that there are still some rough edges, and while trying out the Pydantic V2 beta releases you may experience errors.

>     We encourage you to try out the beta releases in a test environment and not in production.

Reviewer 2: In a couple of lines above it says: "We believe Pydantic V2 is ready for your production applications." 🤔

    While we do not currently intend to make changes to the API, it's still possible that we
    will discover problems requiring breaking changes.

If you do encounter any issues, please [create an issue in GitHub](https://github.com/pydantic/pydantic/issues) using
the `bug V2` label. This will help us to actively monitor and track errors, and to continue to improve the library's
performance.

## Migration guide

### Changes to `pydantic.BaseModel`

Various method names have been changed; all non-deprecated `BaseModel` methods now have names matching either the
format `model_.*` or `__.*pydantic.*__`. Where possible, we have retained the deprecated methods with their old names
to help ease migration, but calling them will emit `DeprecationWarning`s.

| Pydantic V2  | Pydantic V1 |
| ------------ | ----------- |
| `model_dump()` | `dict()` |
| `model_dump_json()` | `json()` |
| `model_copy()` | `copy()` |
| `model_json_schema()` | `json_schema()` |
| `model_construct()` | `construct()` |
| `model_validate()` | `parse_obj()` |
| `__pydantic_validator__` | `__validators__` |
| `model_fields` | `__fields__` |
| `__signature__` | `__signature__` |

> | `__class_vars__` | `__class_vars__` |

Reviewer 2: Those two are the same.

| `__pydantic_private__` | `__private_attributes__` |

> * Some of the built-in data-loading functionality has been slated for removal. In particular,

Reviewer 2: First time I read "slated"... 🤣

>     `parse_raw` and `parse_file` are now deprecated. In Prefect V2, `model_validate_json` works like `parse_raw`. Otherwise, you should load the data and then pass it to `model_validate`.

Reviewer 2: Almost! 👀 
```suggestion
    `parse_raw` and `parse_file` are now deprecated. In Pydantic V2, `model_validate_json` works like `parse_raw`. Otherwise, you should load the data and then pass it to `model_validate`.
```

> * The `from_orm` method has been removed; you can now just use `model_validate` (equivalent to `parse_obj` from

Reviewer 2: I can still see it in the code. 🤔 

```suggestion
* The `from_orm` method has been deprecated; you can now just use `model_validate` (equivalent to `parse_obj` from
```

  Pydantic V1) to achieve something similar, as long as you've set `from_attributes=True` in the model config.
* The `__eq__` method has changed for models.
    * Models can only be equal to other `BaseModel` instances.
    * For two model instances to be equal, they must have the same:
        * Type (or, in the case of generic models, non-parametrized generic origin type)
        * Field values
        * Extra values (only relevant when `model_config['extra'] == 'allow'`)

>         * Private attribute values; models with different values of private attributes are no longer equal.

Reviewer 2: Does this still applies with the latest changes on `PrivateAttr`? Just asking.

Reviewer: @dmontagu could you address this comment? I've taken care of the rest

        * Models are no longer equal to the dicts containing their data.
        * Non-generic models of different types are never equal.
        * Generic models with different origin types are never equal. We don't require *exact* type equality so that,
            for example, instances of `MyGenericModel[Any]` could be equal to instances of `MyGenericModel[int]`.
* We have replaced the use of the `__root__` field to specify a "custom root model" with a new type called
    [`RootModel`](usage/models.md#custom-root-types) which is intended to replace the functionality of using a field
    called `__root__` in Pydantic V1.
* We have significantly expanded Pydantic's capabilities related to customizing serialization. In particular, we have
    added the [`@field_serializer`](api/functional_serializers.md#pydantic.functional_serializers.field_serializer),
    [`@model_serializer`](api/functional_serializers.md#pydantic.functional_serializers.model_serializer), and
    [`@computed_field`](api/fields.md#pydantic.fields.computed_field) decorators, which each address various

>     shortcomings from Pydantic V1.

Reviewer 2: It would be more useful to link the documentation for serializers - I know we don't have it yet.

    * Due to performance overhead and implementation complexity, we have now removed support for specifying
        `json_encoders` in the model config. This functionality was originally added for the purpose of achieving custom
        serialization logic, and we think the new serialization decorators are a better choice in most common scenarios.
        However, if your usage of `json_encoders` is not compatible with the new serialization decorators,

>         please create a GitHub issue letting us know.

Reviewer 2: We should link the `Custom Types` page here, or at least mention how to use `__get_pydantic_json_schema_schema__`.

* `GetterDict` has been removed as it was just an implementation detail of `orm_mode`, which has been removed.

> * `__fields__` in Pydantic V1 has moved to `model_fields` in Pydantic V2.
> * `__validators__` in Pydantic V1 has moved to `__pydantic_validator__` in Pydantic V2.

Reviewer 2: Those two are already in the table above, and they are not adding any new information.

```suggestion
```

### Changes to `pydantic.generics.GenericModel`

The `pydantic.generics.GenericModel` class is no longer necessary, and has been removed. Instead, you can now

> create generic BaseModel subclasses by just adding `Generic` as a parent class on a `BaseModel` subclass directly.

Reviewer 2: ```suggestion
create generic `BaseModel` subclasses by just adding `Generic` as a parent class on a `BaseModel` subclass directly.
```

This looks like `class MyGenericModel(BaseModel, Generic[T]): ...`.

While it may not raise an error, we strongly advise against using _parametrized_ generics in `isinstance` checks.

  * For example, you should not do `isinstance(my_model, MyGenericModel[int])`.
    However, it is fine to do `isinstance(my_model, MyGenericModel)`. (Note that for standard generics, it would raise
    an error to do a subclass check with a parameterized generic.)
  * If you need to perform `isinstance` checks against parametrized generics, you can do this by subclassing the
    parametrized generic class. This looks like `class MyIntModel(MyGenericModel[int]): ...` and
    `isinstance(my_model, MyIntModel)`.

Find more information in the [Generic models](usage/models.md#generic-models) documentation.

### Changes to `pydantic.Field`

`Field` no longer supports arbitrary keyword arguments to be added to the JSON schema. Instead, any extra
data you want to add to the JSON schema should be passed as a dictionary to the `json_schema_extra` keyword argument.

The following properties have been removed from `Field`:

- `const`
- `min_items`

> - `max_items`

Reviewer 2: Those became `min_length` and `max_length`.

- `unique_items`

> - `allow_mutation`

Reviewer 2: Instead use `frozen`.

> - `regex`

Reviewer 2: Became `pattern`.

    * [TODO: Add link to TypeAdapter documentation. You can find example usage in `tests/test_type_adapter.py`.]
* In Pydantic V1, if you used a vanilla (i.e., non-Pydantic) dataclass as a field, the config of the parent type would
  be used as though it was the config for the dataclass itself as well. In Pydantic V2, this is no longer the case.
    * [TODO: Need to specify how to override the config used for vanilla dataclass; possibly need to add functionality?]

### Changes to config

* In Pydantic V2, to specify config on a model, you should set a class attribute called `model_config` to be a dict
  with the key/value pairs you want to be used as the config. The Pydantic V1 behavior to create a class called `Config`
  in the namespace of the parent `BaseModel` subclass is now deprecated.
* The following config settings have been removed:
    * `allow_mutation`.
    * `error_msg_templates`.
    * `fields` — this was the source of various bugs, so has been removed.
      You should be able to use `Annotated` on fields to modify them as desired.
    * `getter_dict` — `orm_mode` has been removed, and this implementation detail is no longer necessary.

>     * `schema_extra` — you should now use the `json_schema_extra` keyword argument to `pydantic.Field`.

Reviewer 2: What's the difference between `schema_extra` and `json_schema_extra`? If it's a renaming, it should be on the next bullet list, not this one. 🤔

>     * `schema_extra` — you should now use the `json_schema_extra` keyword argument to `pydantic.Field`.

Reviewer 2: This should go to renaming I guess?

Reviewer: I don't think it's technically renaming because we moved the functionality to a completely different class.

Reviewer 2: We do support `json_schema_extra` on `ConfigDict`, so I don't think the sentence is accurate. 🤔

    * `smart_union`.
    * `underscore_attrs_are_private` — the Pydantic V2 behavior is now the same as if this was always set
      to `True` in Pydantic V1.

* The following config settings have been renamed:
    * `allow_population_by_field_name` → `populate_by_name`
    * `anystr_lower` → `str_to_lower`
    * `anystr_strip_whitespace` → `str_strip_whitespace`
    * `anystr_upper` → `str_to_upper`
    * `keep_untouched` → `ignored_types`
    * `max_anystr_length` → `str_max_length`
    * `min_anystr_length` → `str_min_length`
    * `orm_mode` → `from_attributes`

>     * `validate_all` → `validate_default`

Reviewer 2: Maybe add a link to the `Model Config` documentation page?

`GenerateJsonSchema` class as necessary for your application.

### `BaseSettings` has moved to `pydantic-settings`

[`BaseSettings`](api/pydantic_settings.md#pydantic_settings.BaseSettings), the base object for Pydantic [settings management](usage/pydantic_settings.md), has been moved to a separate package, [`pydantic-settings`](https://github.com/pydantic/pydantic-settings), which is now a dependency of Pydantic.

## Moved in Pydantic V2

| Pydantic V1 | Pydantic V2 |
| --- | --- |
| `pydantic.utils.version_info` | `pydantic.version.version_info` |
| `pydantic.error_wrappers.ValidationError` | `pydantic.ValidationError` |
| `pydantic.utils.to_camel` | `pydantic.alias_generators.to_pascal` |
| `pydantic.utils.to_lower_camel` | `pydantic.alias_generators.to_camel` |

In addition, the following special-use types have been moved to the [Pydantic Extra Types](https://github.com/pydantic/pydantic-extra-types) project and may be installed as an separate option when needed.

* [Color Types](usage/types/extra_types/color_types.md)
* [Payment Card Numbers](usage/types/extra_types/payment_cards.md)

> * [Phone Numbers](usage/types/extra_types/phone_numbers.md)

Reviewer: ```suggestion
```

> * [Routing Numbers](usage/types/extra_types/routing_numbers.md)

Reviewer: ```suggestion
```

    """
```

### Introduction of `TypeAdapter`

Pydantic V1 had weak support for validating or serializing non-`BaseModel` types.

To work with them, you had to either create a "root" model or use the utility functions in `pydantic.tools`
(namely, `parse_obj_as` and `schema_of`).

In Pydantic V2 this is _a lot_ easier: the `TypeAdapter` class lets you create an object with methods for validating,
serializing, and producing JSON schemas for arbitrary types. This serves as a complete replacement for `parse_obj_as`
and `schema_of` (which are now deprecated), and also covers some of the use cases of "root" models
(`RootModel`, discussed above, covers the others).

```python
from typing import List

from pydantic import TypeAdapter

adapter = TypeAdapter(List[int])
assert adapter.validate_python(['1', '2', '3']) == [1, 2, 3]
print(adapter.json_schema())
#> {'items': {'type': 'integer'}, 'type': 'array'}
```

Due to limitations of inferring generic types with common type checkers, to get proper typing in some scenarios, you
may need to explicitly specify the generic parameter:

> ```python
> from pydantic import TypeAdapter
> 
> adapter: TypeAdapter[str | int] = TypeAdapter(str | int)
> ...
> ```

Reviewer 3: ```suggestion
```python
from __future__ import annotations

from pydantic import TypeAdapter

adapter: TypeAdapter[str | int] = TypeAdapter(str | int)
...
```
```

The main changes are:

* `__get_validators__` should be replaced with `__get_pydantic_core_schema__`
    * [TODO: Add link to documentation of `__get_pydantic_core_schema__`]
* `__modify_schema__` becomes `__get_pydantic_json_schema__`
    * [TODO: Add link to documentation of `__get_pydantic_json_schema__`]

Additionally, you can use `typing.Annotated` to modify or provide the `__get_pydantic_core_schema__` and
`__get_pydantic_json_schema__` functions of a type by annotating it, rather than modifying the type itself.
This provides a powerful and flexible mechanism for integrating third-party types with Pydantic, and in some cases
may help you remove hacks from Pydantic V1 introduced to work around the limitations for custom types.

See [Custom Data Types](usage/types/custom.md) for more information.

### Changes to JSON schema generation

We received many requests over the years to make changes to the JSON schemas that pydantic generates.

In Pydantic V2, we have tried to address many of the common requests:

* The JSON schema for `Optional` fields now indicates that the value `null` is allowed.
* The `Decimal` type is now exposed in JSON schema (and serialized) as a string.

> * The JSON schema no longer preserves named tuples as named tuples.

Reviewer 2: Is this supposed to be:
```suggestion
* The JSON schema no longer preserves namedtuples as namedtuples.
```
?

- `pydantic.typing.all_literal_values`
- `pydantic.typing.display_as_type`
- `pydantic.typing.get_all_type_hints`
- `pydantic.typing.get_args`
- `pydantic.typing.get_origin`
- `pydantic.typing.get_sub_types`
- `pydantic.typing.is_callable_type`
- `pydantic.typing.is_classvar`
- `pydantic.typing.is_finalvar`
- `pydantic.typing.is_literal_type`
- `pydantic.typing.is_namedtuple`
- `pydantic.typing.is_new_type`
- `pydantic.typing.is_none_type`
- `pydantic.typing.is_typeddict`
- `pydantic.typing.is_typeddict_special`
- `pydantic.typing.is_union`
- `pydantic.typing.new_type_supertype`
- `pydantic.typing.resolve_annotations`
- `pydantic.typing.typing_base`
- `pydantic.typing.update_field_forward_refs`
- `pydantic.typing.update_model_forward_refs`
- `pydantic.utils.ClassAttribute`
- `pydantic.utils.DUNDER_ATTRIBUTES`
- `pydantic.utils.PyObjectStr`
- `pydantic.utils.ValueItems`
- `pydantic.utils.almost_equal_floats`
- `pydantic.utils.get_discriminator_alias_and_values`
- `pydantic.utils.get_model`
- `pydantic.utils.get_unique_discriminator_alias`
- `pydantic.utils.in_ipython`
- `pydantic.utils.is_valid_identifier`
- `pydantic.utils.path_type`
- `pydantic.utils.validate_field_name`

> - `pydantic.validate_model`

Reviewer 2: Should we mention `pydantic-extra-types` for `Color` and `PaymentCardNumber`?

| `pydantic.utils.to_camel` | `pydantic.alias_generators.to_pascal` |
| `pydantic.utils.to_lower_camel` | `pydantic.alias_generators.to_camel` |

In addition, the following special-use types have been moved to the [Pydantic Extra Types](https://github.com/pydantic/pydantic-extra-types) project and may be installed as an separate option when needed.

* [Color Types](usage/types/extra_types/color_types.md)
* [Payment Card Numbers](usage/types/extra_types/payment_cards.md)
* [Phone Numbers](usage/types/extra_types/phone_numbers.md)

> * [Routing Numbers](usage/types/extra_types/routing_numbers.md)

Reviewer 2: Those were not in V1, no need to add them here.
